adal/src/main/java/com/microsoft/aad/adal/TokenCacheItem.java

```diff
@@ -122,122 +129,259 @@ public static TokenCacheItem createRegularTokenCacheItem(final String authority,
+     * @return the refresh token string.
+     */
     public String getRefreshToken() {
         return mRefreshtoken;
     }
 
+    /**
+     * Set the fresh token string.
+     * 
+     * @param refreshToken the refresh token string.
+     */
     public void setRefreshToken(String refreshToken) {
         this.mRefreshtoken = refreshToken;
     }
 
+    /**
+     * Get the expire date.
+     * 
+     * @return the time the token get expired.
+     */
     public Date getExpiresOn() {
-        return mExpiresOn;
+        return mExpiresOn != null ? new Date(mExpiresOn.getTime()) : null;
     }
 
+    /**
+     * Set the expire date.
+     * 
+     * @param expiresOn the expire time.
+     */
     public void setExpiresOn(Date expiresOn) {
-        this.mExpiresOn = expiresOn;
+        this.mExpiresOn = (Date) expiresOn.clone();
     }
 
+    /**
+     * Get the multi-resource refresh token flag.
+     * 
+     * @return true if the token is a multi-resource refresh token, else return false.
+     */
     public boolean getIsMultiResourceRefreshToken() {
         return mIsMultiResourceRefreshToken;
     }
 
+    /**
+     * Set the multi-resource refresh token flag.
+     * 
+     * @param isMultiResourceRefreshToken true if the token is a multi-resource refresh token.
+     */
     public void setIsMultiResourceRefreshToken(boolean isMultiResourceRefreshToken) {
         this.mIsMultiResourceRefreshToken = isMultiResourceRefreshToken;
     }
 
+    /**
+     * Get tenant identifier.
+     * 
+     * @return the tenant identifier string.
+     */
     public String getTenantId() {
         return mTenantId;
     }
 
+    /**
+     * Set tenant identifier.
+     * 
+     * @param tenantId the tenant identifier string.
+     */
     public void setTenantId(String tenantId) {
         this.mTenantId = tenantId;
     }
 
+    /**
+     * Get raw ID token.
+     * 
+     * @return raw ID token string.
+     */
     public String getRawIdToken() {
         return mRawIdToken;
     }
 
+    /**
+     * Set raw ID token.
+     * 
+     * @param rawIdToken raw ID token string.
+     */
     public void setRawIdToken(String rawIdToken) {
         this.mRawIdToken = rawIdToken;
     }
-    
+
+    /**
+     * Get family client identifier.
+     * 
+     * @return the family client ID string.
+     */
     public final String getFamilyClientId() {
         return mFamilyClientId;
     }
-    
+
+    /**
+     * Set family client identifier.
+     * 
+     * @param familyClientId the family client ID string.
+     */
     public final void setFamilyClientId(final String familyClientId) {
         this.mFamilyClientId = familyClientId;
     }
 
+    /**
+     * Set the extended expired time.
+     * 
+     * @param extendedExpiresOn extended expired date.
+     */
     public final void setExtendedExpiresOn(final Date extendedExpiresOn) {
-        mExtendedExpiresOn = extendedExpiresOn;
+        mExtendedExpiresOn = (Date) extendedExpiresOn.clone();
```

Comment:
If you want to create a new date, i would rather you do new Date(extendedExpirsOn.getTime()). Also, i'm just curious, why you change all the set for Date into this?